### DISCUSSION TITLE
Pass a custom network namespace id where docker should run the containers

### DISCUSSION THREAD

**Discussion by rawahars** [moby/moby#45272] [Q&A]
Is there a way we can pass in a custom network namespace ID where the docker container can be ran?

Alternatively, I am interested in finding the code path where we call OS specific calls for creating network namespaces. For example, when we use `--net none`, we get an empty network namespace for the container. From where do we create this network namespace in the code?

**Comment by akerouanton** [comment]
Hi @rawahars,

No, unfortunately there's no way to specify an "external" netns. What are you trying to achieve? Can you describe your use-case please? That'd help us determine whether this deserves a feature request.

As for how network namespaces are created, libnetwork's controller is in charge of this. You can take a look at the [`NewSandbox` method implemented by this controller](https://github.com/moby/moby/blob/e8a79114b8c1782d5539421c9a6a1cd6fc5dfa65/libnetwork/controller.go#L846). You might also be interested by how links (eg. veth pairs) are created and moved into the appropriate netns. This is handled by netdrivers, more specifically by their `Join` method. You can take a look at the `Join` method of the bridge driver ([see here](https://github.com/moby/moby/blob/e8a79114b8c1782d5539421c9a6a1cd6fc5dfa65/libnetwork/drivers/bridge/bridge.go#L1203)).

**Reply by rawahars** [reply]
Thank you @akerouanton for your response.

My use case if around supporting `host` network mode for `Windows containers`. Presently, it is not supported with docker using `--net host`.
However, I was able to verify that we can accomplish the same using a [namespace](https://github.com/microsoft/hcsshim/blob/62680e0b85326a67506094fc0f89ceac239b8820/hcn/hcnnamespace.go#L48) of type `HostDefault`. 

We can do it easily with `containerd` wherein we-
- First create a host network namespace using `hcsshim`
- Pass the namespace ID when we launch the containers

**Reply by akerouanton** [reply]
I think we don't want to give users a way to specify their own network namespaces because it's really a low-level thing and how the netns is set up could badly influence docker.

However, adding `--net=host` to Windows is probably something we want to do. I did a quick check and I can't find any issue with such feature request. Can you create one please?

**Reply by rawahars** [reply]
@akerouanton Thanks for your response. I have created an issue here- https://github.com/moby/moby/issues/45307

I will be happy to contribute the changes if you can guide me about the workflow which needs modification. Also, are there any Microsoft maintainers we can work with on this?

CC: @jterry75 @ofiliz

**Reply by thaJeztah** [reply]
I think both of them are "same team, different office" now, but we can look if there's Microsoft maintainers to help with review if needed 👍
